File: src/main/java/eu/insertcode/wordgames/games/WordGame.java
package eu.insertcode.wordgames.games;

import org.bukkit.Bukkit;
import org.bukkit.ChatColor;
import org.bukkit.entity.Player;
import org.bukkit.event.player.AsyncPlayerChatEvent;

import eu.insertcode.wordgames.Main;
import eu.insertcode.wordgames.Permission;

import static org.bukkit.ChatColor.translateAlternateColorCodes;

public abstract class WordGame {
	final Reward reward;
	final Main plugin;
	String showedWord,
			wordToType;
	
	WordGame(Main instance, String wordToType, Reward reward) {
		plugin = instance;
		this.wordToType = wordToType;
		this.reward = reward;
		this.showedWord = wordToType;
	}
	
	public void endGame() {
		plugin.removeGame(this);
	}
	
	void sendWinnerMessage(Player winner) {
		String[] messages = Main.getMessages("games.gameWon");
		for (String message : messages) {
			message = formatGameMessage(message, wordToType).replace("{player}", winner.getName());
			Bukkit.broadcastMessage(translateAlternateColorCodes('&', message));
		}
	}
	
	public abstract Permission getPlayPermission();
	
	abstract String getMessageConfigPath();
	
	public String[] getGameMessages() {
		String[] messages = Main.getMessages(getMessageConfigPath());
		for (int i = 0; i < messages.length; i++) {
			messages[i] = formatGameMessage(messages[i], showedWord);
			messages[i] = translateAlternateColorCodes('&', messages[i]);
			
		}
		return messages;
	}
	
	void sendGameMessage() {
		String[] messages = getGameMessages();
		for (String message : messages) {
			Bukkit.broadcastMessage(message);
		}
	}
	
	String formatGameMessage(String message, String word) {
		return message.replace("{word}", word)
				.replace("{amount}", "" + reward.getAmount())
				.replace("{reward}", reward.getReward());
	}
	
	public void onPlayerChat(AsyncPlayerChatEvent e) {
		Bukkit.getScheduler().runTask(plugin, () -> {
			Player p = e.getPlayer();
			//If the player types the correct word.
			String message = ChatColor.stripColor(e.getMessage()).trim();
			if (message.equalsIgnoreCase(wordToType)) {
				if (!Permission.PLAY_ALL.forPlayer(p, getPlayPermission())) {
					for (String msg : Main.getColouredMessages("error.noPlayPermissions"))
						p.sendMessage(msg);
					return;
				}
				
				String command = plugin.getConfig().getString("gameOptions.rewardCommandSyntax");
				command = command.replace("{username}", p.getName()).replace("{reward}", reward.getReward()).replace("{amount}", "" + reward.getAmount());
				Bukkit.dispatchCommand(plugin.getServer().getConsoleSender(), command);
				
				sendWinnerMessage(p);
				endGame();
			}
		});
	}
	
	public static class Reward {
		private final int amount;
		private final String reward;
		
		public Reward(int amount, String reward) {
			this.amount = amount;
			this.reward = reward;
		}
		
		int getAmount() {
			return amount;
		}
		
		String getReward() {
			return reward;
		}
	}
}

File: src/main/java/eu/insertcode/wordgames/games/WordGame.kt
package eu.insertcode.wordgames.games

import eu.insertcode.wordgames.Main
import eu.insertcode.wordgames.Permission
import eu.insertcode.wordgames.config.Config
import eu.insertcode.wordgames.config.Messages
import eu.insertcode.wordgames.message.MessageHandler.getMessages
import eu.insertcode.wordgames.message.MessageHandler.sendMessage
import org.bukkit.Bukkit
import org.bukkit.ChatColor
import org.bukkit.entity.Player
import org.bukkit.event.player.AsyncPlayerChatEvent

abstract class WordGame(
    val plugin: Main,
    var wordToType: String,
    val reward: Reward,
    var showedWord: String = wordToType
) {

    open fun endGame() {
        plugin.removeGame(this)
    }

    open fun sendWinnerMessage(winner: Player) {
        for (message in getMessages(Messages.Games.gameWon)) {
            Bukkit.broadcastMessage(ChatColor.translateAlternateColorCodes('&',
                formatGameMessage(message, wordToType).replace("{player}", winner.displayName)
            ))
        }
    }

    abstract val playPermission: Permission
    abstract val messageConfigPath: String
    open val gameMessages: List<String>
        get() = getMessages(messageConfigPath).map {
            ChatColor.translateAlternateColorCodes('&', formatGameMessage(it, showedWord))
        }

    open fun sendGameMessage() {
        for (message in gameMessages) {
            Bukkit.broadcastMessage(message)
        }
    }

    open fun formatGameMessage(message: String, word: String) =
        message.replace("{word}", word)
            .replace("{amount}", reward.amount.toString())
            .replace("{reward}", reward.reward)

    open fun onPlayerChat(e: AsyncPlayerChatEvent) {
        Bukkit.getScheduler().runTask(plugin, Runnable {
            val p = e.player
            //If the player types the correct word.
            val message = ChatColor.stripColor(e.message)!!.trim { it <= ' ' }
            if (message.equals(wordToType, ignoreCase = true)) {
                if (!Permission.PLAY_ALL.forPlayer(p, playPermission)) {
                    sendMessage(p, Messages.Error.noPlayPermissions, true)
                    return@Runnable
                }

                val command = Config.GameOptions.rewardCommandSyntax.replace("{username}", p.name).replace("{reward}", reward.reward).replace("{amount}", "" + reward.amount)
                Bukkit.dispatchCommand(plugin.server.consoleSender, command)
                sendWinnerMessage(p)
                endGame()
            }
        })
    }
}